Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -1257,6 +1262,20 @@ public Task<Result<byte[]>> GetBlobAsync(OperationContext context, ContentHash h
             return GlobalStore.GetBlobAsync(context, hash);
         }
 
+        private async void OnContentLocationDatabaseInvalidationAsync(object sender, ContentLocationDatabase.DatabaseInvalidatedEventArgs e)
+        {
+            // If multiple threads fail at the same time (i.e. corruption error), this cheaply deduplicates the
+            // restores, avoids redundant logging.
+            using (_databaseInvalidationGate.AcquireSemaphore(0))

```

Comment:
I'm curious, have you seen more then one logging at the same time? Because ProcessStateAsync will take care of concurrency any ways... And I Can't see how you actually reducing the amount of logging by "serializing" the body of this using block.

Reply:
Yes, I have seen this. An example situation is as follows:

- Imagine K threads running in parallel with .Use()
- DB gets corrupted, threads perform some DB operation after the corruption, cause a throw
- All threads simultaneously disable the store (there is no semaphore in that code)
- Hence, the invalidation handler is called K times, and possibly spaced out

This semaphore is there to ensure that we only log an error once per machine. Let me know if you want a query that logs this example.

---
In reply to: [327866401](https://github.com/microsoft/BuildXL/pull/923#discussion_r327866401) [](ancestors = 327866401)